lib/core/util.c

```diff
@@ -944,3 +944,119 @@ void h2o_cleanup_thread(void)
+    int fd = 0;
+    u_int32_t now = (u_int32_t)time(NULL);
+
+    // only try every second
+    if ((h2o_tracing_map_last_attempt - now) == 0) {
+        return 0;
+    }
+    __sync_bool_compare_and_swap(&h2o_tracing_map_last_attempt, h2o_tracing_map_last_attempt, now);
+
+    memset(&attr, 0, sizeof(attr));
+    attr.pathname = (__u64)(unsigned long)&h2o_ebpf_map_path[0];
+
+    fd = syscall(__NR_bpf, BPF_OBJ_GET, &attr, sizeof(attr));
+    if (fd <= 0)
+      return 0;
+
+    // try set map fd if not set before
+    // if fd was set beforehand, don't set the new fd, and close it
+    if (!__sync_bool_compare_and_swap(&h2o_tracing_map_fd, 0, fd))
+        close(fd);
+
+    return h2o_tracing_map_fd;
+}
+
+inline int lookup_map(const void *key, const void *value)
+{
+    union bpf_attr attr;
+
+    memset(&attr, 0, sizeof(attr));
+    attr.map_fd = h2o_tracing_map_fd;
+    attr.key = (__u64)(unsigned long)key;
+    attr.value = (__u64)(unsigned long)value;
+
+    return syscall(__NR_bpf, BPF_MAP_LOOKUP_ELEM, &attr, sizeof(attr));
+}
+
+inline void read_ip_port(struct sockaddr *sa, void *ip, u_int16_t *port)
+{
+    if (sa->sa_family == AF_INET) {
+        struct sockaddr_in *sin = (void *)sa;
+        memcpy(ip, &sin->sin_addr, sizeof(sin->sin_addr));
+        *port = sin->sin_port;
+    } else if (sa->sa_family == AF_INET6) {
+        struct sockaddr_in6 *sin = (void *)sa;
+        memcpy(ip, &sin->sin6_addr, sizeof(sin->sin6_addr));
+        *port = sin->sin6_port;
+    }
+}
+
+int h2o_tracing_is_sock_traced(h2o_socket_t *sock)
+{
+    // check for tracing enablement
+    if (H2O_UNLIKELY(!H2O_H2O_CONN_TRACING_ENABLED())) {
+
+        // cleanup opened map if it was opened
+        if (H2O_UNLIKELY(h2o_tracing_map_fd > 0)) {
+            close(h2o_tracing_map_fd);
+            __sync_fetch_and_and(&h2o_tracing_map_fd, 0);
+        }
+
+        return 0;
+    }
+
+    // try open map if not opened
+    if (h2o_tracing_map_fd <= 0 && open_map(h2o_tracing_map_fd) <= 0)
+        return 1; // map can't be opened, fallback accepting probe
+
+    // define key/vals - we are only interrested in presence of the key, discard values
+    struct h2o_ebpf_map_key_t key;
+    memset(&key, 0, sizeof(key));
+    void *vals = NULL;
+
+    // get sock/peer ip/ports
+    struct sockaddr_storage loc, rem;
+    h2o_socket_getsockname(sock, (void *)&loc);
+    h2o_socket_getpeername(sock, (void *)&rem);
+
+    // read ip/ports, put parsed val into key structure
+    read_ip_port((void *)&loc, &key.ipa, &key.porta);
+    read_ip_port((void *)&rem, &key.ipb, &key.portb);
+
+    // read ip family
+    key.family = loc.ss_family == AF_INET6 ? 6 : 4;
+
+    // lookup map for our key
+    if (lookup_map(&key, &vals) == -1)
+        return 0; // key not in map, or errored - return 0
+
+    H2O_H2O_CONN_TRACING(); // dummy call - only used for checking enablement
+    return 1;
+}
+#else
+int h2o_tracing_is_sock_traced(h2o_socket_t *sock)
+{
+    return 1;
+}
+#endif
```

Comment:
Please move the socket-level functions to socket.c.

Our general rule is to not have dependency from files under lib/common to lib/core.